src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs

```diff
@@ -76,21 +75,50 @@ public class Arguments : UnsupervisedLearnerInputBaseWithWeight
         private readonly int _k;
 
         private readonly int _maxIterations; // max number of iterations to train
-        private readonly Float _convergenceThreshold; // convergence thresholds
+        private readonly float _convergenceThreshold; // convergence thresholds
 
         private readonly long _accelMemBudgetMb;
         private readonly InitAlgorithm _initAlgorithm;
         private readonly int _numThreads;
+        private readonly string _featureColumn;
 
         public override TrainerInfo Info { get; }
         public override PredictionKind PredictionKind => PredictionKind.Clustering;
 
-        public KMeansPlusPlusTrainer(IHostEnvironment env, Arguments args)
-            : base(env, LoadNameValue)
+        /// <summary>
+        /// Initializes a new instance of <see cref="KMeansPlusPlusTrainer"/>
+        /// </summary>
+        /// <param name="env">The private instance of <see cref="IHostEnvironment"/>.</param>
+        /// <param name="featureColumn">The name of the feature column.</param>
+        /// <param name="weightColumn">The name for the column containing the initial weight.</param>
+        /// <param name="advancedSettings">A delegate to apply all the advanced arguments to the algorithm.</param>
+        public KMeansPlusPlusTrainer(IHostEnvironment env, string featureColumn, string weightColumn = null, Action<Arguments> advancedSettings = null)
+            : this(env, null, featureColumn, weightColumn, advancedSettings)
+        {
+
+        }
+
+        internal KMeansPlusPlusTrainer(IHostEnvironment env, Arguments args)
+            : this(env, args, args.FeatureColumn, args.WeightColumn, null)
+        {
+
+        }
+
+        private KMeansPlusPlusTrainer(IHostEnvironment env, Arguments args, string featureColumn, string weightColumn, Action<Arguments> advancedSettings = null)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(LoadNameValue), TrainerUtils.MakeR4VecFeature(featureColumn), null, TrainerUtils.MakeR4ScalarWeightColumn(weightColumn))
         {
-            Host.CheckValue(args, nameof(args));
+            if (args == null)
+                args = new Arguments();
```

Comment:
The net effect of this change is that the internal constructor is now tolerant to passing in `null` arguments.. Even though the constructor is internal I'd prefer to keep it clean, and maintain the invariants we have everywhere else. Note the `Host.CheckValue(args, nameof(args));` check that was deleted.

If you still had the check, and just passed in `new Arguments` in the above constructor, that would have had no more lines of code.

Reply:
appreciate pointing this out. 

---
In reply to: [220271405](https://github.com/dotnet/machinelearning/pull/979#discussion_r220271405) [](ancestors = 220271405)